pex/pex_builder.py

```diff
@@ -626,18 +651,138 @@ def freeze(self, bytecode_compile=True):
+    ):
+        # type: (...) -> None
+        """Package the PEX application.
+
+        By default, the PEX is packaged as a zipapp for ease of shipping as a single file, but it
+        can also be packaged in spread mode for efficiency of syncing over the network
+        incrementally.
 
-        :param filename: The filename where the PEX should be stored.
+        :param path: The path where the PEX should be stored.
         :param bytecode_compile: If True, precompile .py files into .pyc files.
         :param deterministic_timestamp: If True, will use our hardcoded time for zipfile timestamps.
+        :param spread: If True, will build the PEX in spread directory mode.
 
         If the PEXBuilder is not yet frozen, it will be frozen by ``build``.  This renders the
         PEXBuilder immutable.
         """
         if not self._frozen:
             self.freeze(bytecode_compile=bytecode_compile)
+        if spread:
+            self._build_spreadapp(dirname=path, deterministic_timestamp=deterministic_timestamp)
+        else:
+            self._build_zipapp(filename=path, deterministic_timestamp=deterministic_timestamp)
+
+    def _build_spreadapp(
+        self,
+        dirname,  # type: str
+        deterministic_timestamp=False,  # type: bool
+    ):
+        # type: (...) -> None
+
+        pex_info = self._pex_info
+
+        safe_rmtree(dirname)
+
+        # N.B.: We want an atomic directory, but we don't expect a user to race themselves building
+        # to a single non-PEX_ROOT user-requested output path; so we don't grab an exclusive lock
+        # and dirty the target directory with a `.lck` file.
+        with atomic_directory(dirname, exclusive=False) as atomic_dir:
+            if not atomic_dir.is_finalized:
+                sources = []  # type: List[str]
+                spreads = []  # type: List[Spread]
+
+                work_dir = atomic_dir.work_dir
+
+                # Include user sources, PEX-INFO and __main__ as loose files in src/.
+                for fileset in "source", "resource", "executable", "main", "manifest":
+                    for f in self._chroot.filesets.get(fileset, ()):
+                        dest = os.path.join(work_dir, "src", f)
+                        safe_mkdir(os.path.dirname(dest))
+                        safe_copy(os.path.join(self._chroot.chroot, f), dest)
+                        sources.append(f)
+
+                # Zip up the bootstrap which is constant for a given version of Pex.
+                bootstrap_root = os.path.join(self._chroot.chroot, pex_info.bootstrap)
+                bootstrap_fingerprint = CacheHelper.dir_hash(bootstrap_root)
+                cached_bootstrap_zip_dir = os.path.join(
+                    pex_info.pex_root, "bootstrap_zips", bootstrap_fingerprint
+                )
+                with atomic_directory(
+                    cached_bootstrap_zip_dir, exclusive=False
+                ) as atomic_bootstrap_zip_dir:
```

Comment:
There's an explanation above for why the destination is not marked exclusive, but should these cache directories be exclusive?

Reply:
No, because they are creating a single zip file. So atomicity is guaranteed without a lock. This is unlike loose files that could get partially pyc compiled.